best-practices/client-side/framework/react/components/media-cropper/src/components/media-cropper/renderers/image.js

```diff
@@ -0,0 +1,68 @@
+import React, { Component } from 'react';
+
+export default class ImageRenderer extends Component {
+    ctx = null;
+
+    get croppedMedia() {
+        this.setState({
+            ...this.state,
+            rendering: true
+        })
+        return new Promise(resolve => {
+            this.refs.canvas.toBlob((blob) => {
+                this.setState({
+                    ...this.state,
+                    rendering: false
+                })
+                resolve(blob);
+            }, 'image/png');
+        });
+    }
+
+    constructor(props) {
+        super(props);
+        this.state = { ...this.props, rendering: false };
+    }
+
+    componentDidMount() {
+        this.updateContext();
+    }
+
+    componentDidUpdate(props, state) {
+        // if we are switching to rendering
+        if (!state.rendering && this.state.rendering) {
+            console.log('freezing state');
+            // freeze the current state
+            this.setState({ ...this.state, ...this.props });
+        }
+        this.renderCanvas();
+    }
```

Comment:
Related to the earlier comments about hooks, the `useEffect` hook is super cool for replacing these different event hooks.